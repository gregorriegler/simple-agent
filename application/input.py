from typing import Callable, List, Optional


class Input:

    def __init__(self, display, escape_detector: Optional[Callable[[], bool]] = None):
        self.display = display
        self._stack: List[str] = []
        self._escape_detector = escape_detector

    def stack(self, message: str):
        self._stack.append(message)

    def has_stacked_messages(self) -> bool:
        return bool(self._stack)

    def read(self) -> str:
        if self._stack:
            return self._stack.pop()
        return self.display.input()

    def escape_requested(self) -> bool:
        if self._escape_detector is None:
            return False
        return self._escape_detector()

    def set_escape_detector(self, escape_detector: Callable[[], bool]):
        self._escape_detector = escape_detector
