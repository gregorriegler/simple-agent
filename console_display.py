from agent import Display
import sys


class ConsoleDisplay(Display):

    def __init__(self, indent_level=0, print_fn=print, agent_type="Agent"):
        self.indent_level = indent_level
        self.print = print_fn
        self.agent_type = agent_type
        self.base_indent = "       " * (indent_level + 1)  # 7 spaces per level
        self.agent_prefix = "       " * indent_level + f"{agent_type}: "

    def assistant_says(self, message):
        lines = str(message).split('\n')
        if lines:

            self.print(f"\n{self.agent_prefix}{lines[0]}")
            for line in lines[1:]:
                self.print(f"{self.base_indent}{line}")

    def tool_about_to_execute(self, parsed_tool):
        indented_tool = self._indent_lines(str(parsed_tool), "       " * self.indent_level)
        self.print(f"\n{indented_tool}")

    def tool_result(self, result):
        result = self._indent_lines(result, self.base_indent)
        self.print(f"\n{result}")

    def input(self):
        prompt = "       " * self.indent_level + "Press Enter to continue or type a message to add: "
        self.print(f"\n{prompt}", file=sys.stderr)
        return input("")

    def exit(self):
        exit_msg = "       " * self.indent_level + "Exiting..."
        self.print(f"\n{exit_msg}")

    def _indent_lines(self, text, prefix="    "):
        lines = str(text).split('\n')
        return '\n'.join(f"{prefix}{line}" if line.strip() else line for line in lines)
