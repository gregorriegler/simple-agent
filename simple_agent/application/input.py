from typing import List

from .user_input import UserInput


class Input:

    def __init__(self, user_input: UserInput):
        self.user_input = user_input
        self._stack: List[str] = []

    def stack(self, message: str):
        self._stack.append(message)

    def has_stacked_messages(self) -> bool:
        return bool(self._stack)

    def read(self) -> str:
        if self._stack:
            return self._stack.pop()
        return self.user_input.read()

    def escape_requested(self) -> bool:
        return self.user_input.escape_requested()
