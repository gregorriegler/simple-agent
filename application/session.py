from dataclasses import dataclass

from application.agent import Agent
from application.llm import Messages
from application.persisted_messages import PersistedMessages
from application.session_storage import SessionStorage
from application.system_prompt import SystemPrompt
from tools import ToolLibrary


@dataclass
class SessionArgs:
    continue_session: bool
    start_message: str | None


def run_session(
    continue_session: bool,
    user_input,
    display,
    session_storage: SessionStorage,
    chat,
    system_prompt: SystemPrompt
):
    messages = session_storage.load() if continue_session else Messages()
    persisted_messages = PersistedMessages(messages, session_storage)

    if continue_session:
        display.continue_session()
    else:
        display.start_new_session()
    tool_library = ToolLibrary(chat)
    system_prompt = system_prompt()
    agent = Agent(chat, system_prompt, user_input, tool_library, display, session_storage)
    agent.start(persisted_messages)
