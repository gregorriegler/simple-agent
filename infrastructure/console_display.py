from application.display import Display
import sys


class ConsoleDisplay(Display):

    def __init__(self, indent_level=0, agent_name="Agent", input_fn=input, print_fn=print):
        self.indent_level = indent_level
        self.print = print_fn
        self.base_indent = "       " * (indent_level + 1)  # 7 spaces per level
        self.agent_prefix = "       " * indent_level + f"{agent_name}: "
        self.input_fn = input_fn

    def assistant_says(self, message):
        lines = str(message).split('\n')
        if lines:

            self.print(f"\n{self.agent_prefix}{lines[0]}")
            for line in lines[1:]:
                self.print(f"{self.base_indent}{line}")

    def tool_result(self, result):
        lines = str(result).split('\n')
        first_three_lines = '\n'.join(lines[:3])
        if len(lines) > 3:
            first_three_lines += '\n... (truncated)'
        result = self._indent_lines(first_three_lines, self.base_indent)
        self.print(f"\n{result}")

    def continue_session(self):
        self.print("Continuing session")

    def start_new_session(self):
        self.print("Starting new session")

    def input(self) -> str:
        prompt = "       " * self.indent_level + "Press Enter to continue or type a message to add: "
        self.print(f"\n{prompt}", file=sys.stderr)
        return self.input_fn("").strip()

    def exit(self):
        exit_msg = "       " * self.indent_level + "Exiting..."
        self.print(f"\n{exit_msg}")

    def _indent_lines(self, text, prefix="    "):
        lines = str(text).split('\n')
        return '\n'.join(f"{prefix}{line}" if line.strip() else line for line in lines)
