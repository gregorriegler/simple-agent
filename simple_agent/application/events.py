from dataclasses import dataclass
from typing import Any, ClassVar

from simple_agent.application.tool_library import ToolResult


@dataclass
class AgentEvent:
    agent_id: str
    event_name: ClassVar[str] = "agent_event"


@dataclass
class AssistantSaidEvent(AgentEvent):
    event_name: ClassVar[str] = "assistant_said"
    message: str


@dataclass
class ToolCalledEvent(AgentEvent):
    event_name: ClassVar[str] = "tool_called"
    call_id: str
    tool: Any


@dataclass
class ToolResultEvent(AgentEvent):
    event_name: ClassVar[str] = "tool_result"
    call_id: str
    result: ToolResult


@dataclass
class SessionStartedEvent(AgentEvent):
    event_name: ClassVar[str] = "session_started"
    is_continuation: bool


@dataclass
class SessionInterruptedEvent(AgentEvent):
    event_name: ClassVar[str] = "session_interrupted"


@dataclass
class SessionEndedEvent(AgentEvent):
    event_name: ClassVar[str] = "session_ended"


@dataclass
class UserPromptRequestedEvent(AgentEvent):
    event_name: ClassVar[str] = "user_prompt_requested"


@dataclass
class UserPromptedEvent(AgentEvent):
    event_name: ClassVar[str] = "user_prompted"
    input_text: str


@dataclass
class AssistantRespondedEvent(AgentEvent):
    event_name: ClassVar[str] = "assistant_responded"
    response: str


@dataclass
class SubagentFinishedEvent(AgentEvent):
    event_name: ClassVar[str] = "subagent_finished"
    subagent_id: str
