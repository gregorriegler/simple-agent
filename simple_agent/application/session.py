from dataclasses import dataclass

from simple_agent.application.agent import Agent
from simple_agent.application.llm import Messages
from simple_agent.application.persisted_messages import PersistedMessages
from simple_agent.application.session_storage import SessionStorage
from simple_agent.application.system_prompt import SystemPrompt
from simple_agent.tools import ToolLibrary


@dataclass
class SessionArgs:
    continue_session: bool
    start_message: str | None
    show_system_prompt: bool = False


def run_session(
    continue_session: bool,
    user_input,
    display,
    session_storage: SessionStorage,
    chat,
    system_prompt: SystemPrompt
):
    messages = session_storage.load() if continue_session else Messages()
    persisted_messages = PersistedMessages(messages, session_storage)

    if continue_session:
        display.continue_session()
    else:
        display.start_new_session()
    tool_library = ToolLibrary(chat)
    system_prompt = system_prompt()
    agent = Agent(chat, system_prompt, user_input, tool_library, display, session_storage)
    agent.start(persisted_messages)
