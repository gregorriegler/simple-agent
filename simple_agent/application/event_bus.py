from typing import Dict, List, Type
from .event_bus_protocol import EventBus, EventHandler
from .events import AgentEvent


class SimpleEventBus(EventBus):

    def __init__(self):
        self._handlers: Dict[Type[AgentEvent], List[EventHandler]] = {}

    def subscribe(self, event_type: Type[AgentEvent], handler: EventHandler) -> None:
        if event_type not in self._handlers:
            self._handlers[event_type] = []
        self._handlers[event_type].append(handler)

    def unsubscribe(self, event_type: Type[AgentEvent], handler: EventHandler) -> None:
        if event_type in self._handlers:
            try:
                self._handlers[event_type].remove(handler)
                if not self._handlers[event_type]:
                    del self._handlers[event_type]
            except ValueError:
                pass

    def publish(self, event: AgentEvent) -> None:
        event_type = type(event)
        if event_type in self._handlers:
            for handler in self._handlers[event_type]:
                handler(event)
