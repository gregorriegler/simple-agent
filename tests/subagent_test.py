from approvaltests import verify

from simple_agent.application.input import Input
from simple_agent.application.session import run_session
from simple_agent.application.event_bus import SimpleEventBus
from simple_agent.application.events import (
    AssistantRespondedEvent,
    AssistantSaidEvent,
    SessionEndedEvent,
    SessionInterruptedEvent,
    SessionStartedEvent,
    ToolCalledEvent,
    ToolResultEvent,
    UserPromptRequestedEvent,
    UserPromptedEvent,
)
from simple_agent.infrastructure.console.console_display import ConsoleDisplay
from simple_agent.infrastructure.console.console_user_input import ConsoleUserInput
from simple_agent.infrastructure.display_event_handler import DisplayEventHandler
from .event_spy import EventSpy
from .print_spy import IOSpy
from .test_session_storage import SessionStorageStub
from .test_tool_library import ToolLibraryStub


def test_subagent():
    verify_chat(
        ["Create a subagent that says hello", "\n"], [
            "🛠️ subagent coding say hello",
            "hello\n🛠️ complete-task I successfully said hello"
        ]
    )


def test_nested_agent_test():
    verify_chat(
        ["Create a subagent that creates another subagent", "\n"], [
            "🛠️ subagent orchestrator create another subagent",
            "🛠️ subagent coding say nested hello",
            "nested hello\n🛠️ complete-task I successfully said nested hello",
            "🛠️ complete-task I successfully created another subagent",
            "🛠️ complete-task I successfully created a subagent"
        ]
    )


def test_agent_says_after_subagent():
    verify_chat(
        ["Create a subagent that says hello, then say goodbye", "\n"], [
            "🛠️ subagent coding say hello",
            "hello\n🛠️ complete-task I successfully said hello",
            "goodbye"
        ]
    )


def verify_chat(inputs, answers, escape_hits=None, ctrl_c_hits=None):
    llm_stub = create_llm_stub(answers)
    message, *remaining_inputs = inputs
    io_spy = IOSpy(remaining_inputs, escape_hits)
    display = ConsoleDisplay(0, "Agent", io_spy)
    user_input_port = ConsoleUserInput(display.indent_level, io=io_spy)
    user_input = Input(user_input_port)
    user_input.stack(message)
    test_session_storage = SessionStorageStub()
    event_bus = SimpleEventBus()
    display_handler = DisplayEventHandler(display)

    event_spy = EventSpy()
    tracked_events = [
        SessionStartedEvent,
        UserPromptRequestedEvent,
        UserPromptedEvent,
        AssistantSaidEvent,
        AssistantRespondedEvent,
        ToolCalledEvent,
        ToolResultEvent,
        SessionInterruptedEvent,
        SessionEndedEvent,
    ]

    for event_type in tracked_events:
        event_bus.subscribe(event_type, event_spy.record_event)

    event_bus.subscribe(AssistantSaidEvent, display_handler.handle_assistant_said)
    event_bus.subscribe(ToolCalledEvent, display_handler.handle_tool_called)
    event_bus.subscribe(ToolResultEvent, display_handler.handle_tool_result)
    event_bus.subscribe(SessionStartedEvent, display_handler.handle_session_started)
    event_bus.subscribe(SessionInterruptedEvent, display_handler.handle_session_interrupted)
    event_bus.subscribe(SessionEndedEvent, display_handler.handle_session_ended)

    test_tool_library = ToolLibraryStub(
        llm_stub,
        io=io_spy,
        interrupts=[ctrl_c_hits],
        event_bus=event_bus,
        display_event_handler=display_handler
    )

    run_session(
        False,
        "Agent",
        system_prompt_stub,
        user_input,
        llm_stub,
        test_tool_library,
        test_session_storage,
        event_bus
    )

    result = f"# Events\n{event_spy.get_events_as_string()}\n\n# Saved messages:\n{test_session_storage.saved}"
    verify(result)


def create_llm_stub(answer):
    if isinstance(answer, list):
        answer_index = 0

        def llm_stub(system_prompt, messages):
            nonlocal answer_index
            if answer_index < len(answer):
                result = answer[answer_index]
                answer_index += 1
                return result
            if answer:
                return answer[-1]
            return ""

        return llm_stub

    def llm_answer(system_prompt, messages):
        return answer

    return llm_answer


def system_prompt_stub(_):
    return "Test system prompt"
