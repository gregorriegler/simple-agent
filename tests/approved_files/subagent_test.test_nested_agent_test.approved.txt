# Events
Agent:       session_started: False
Agent:         user_prompted: Create a subagent that creates another subagent
Agent:   assistant_responded: 🛠️ subagent orchestrator create another subagent
Agent:           tool_called: Agent::tool_call::1 🛠️ subagent orchestrator create another subagent
Agent/Orchestrator:         user_prompted: create another subagent
Agent/Orchestrator:   assistant_responded: 🛠️ subagent coding say nested hello
Agent/Orchestrator:           tool_called: Agent/Orchestrator::tool_call::1 🛠️ subagent coding say nested hello
Agent/Orchestrator/Coding:         user_prompted: say nested hello
Agent/Orchestrator/Coding:   assistant_responded: nested hello
🛠️ complete-task I successfully said nested hello
Agent/Orchestrator/Coding:        assistant_said: nested hello
Agent/Orchestrator/Coding:           tool_called: Agent/Orchestrator/Coding::tool_call::1 🛠️ complete-task I successfully said nested hello
Agent/Orchestrator/Coding:           tool_result: Agent/Orchestrator/Coding::tool_call::1 I successfully said nested hello
Agent/Orchestrator/Coding: user_prompt_requested: 
Agent/Orchestrator/Coding:         session_ended: 
Agent/Orchestrator:           tool_result: Agent/Orchestrator::tool_call::1 I successfully said nested hello
Agent/Orchestrator:   assistant_responded: 🛠️ complete-task I successfully created another subagent
Agent/Orchestrator:           tool_called: Agent/Orchestrator::tool_call::2 🛠️ complete-task I successfully created another subagent
Agent/Orchestrator:           tool_result: Agent/Orchestrator::tool_call::2 I successfully created another subagent
Agent/Orchestrator: user_prompt_requested: 
Agent/Orchestrator:         session_ended: 
Agent:           tool_result: Agent::tool_call::1 I successfully created another subagent
Agent:   assistant_responded: 🛠️ complete-task I successfully created a subagent
Agent:           tool_called: Agent::tool_call::2 🛠️ complete-task I successfully created a subagent
Agent:           tool_result: Agent::tool_call::2 I successfully created a subagent
Agent: user_prompt_requested: 
Agent:         session_ended: 

# Saved messages:
user: Create a subagent that creates another subagent
assistant: 🛠️ subagent orchestrator create another subagent
user: Result of 🛠️ subagent orchestrator create another subagent
I successfully created another subagent
assistant: 🛠️ complete-task I successfully created a subagent
